OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala

```diff
@@ -0,0 +1,1413 @@
+                        filterMap(compType)(intent.action).keySet.foreach { categories ⇒
+                            if ((intent.categories.isEmpty && categories.isEmpty) || intent.categories.forall(categories.contains)) {
+                                if (intent.iData == undef) {
+                                    overApproximateEdges(intent, filterMap(compType)(intent.action)(categories))
+                                } else {
+                                    testData(intent, filterMap(compType)(intent.action)(categories))
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        //adding explicit intent edges
+        explicitIntents.foreach { i ⇒
+            val caller = i.caller
+            i.calledMethod.foreach { m ⇒
+                edges += ((caller, i.pc) -> m)
+            }
+        }
+    }
+
+    def testData(
+                  intent: implicitIntent,
+                  filters: ListBuffer[intentFilter]
+    ): Unit = {
+
+        for (filter ← filters) {
+            val types = filter.dataTypes
+            val schemes = filter.dataSchemes
+            val data = intent.iData
+            if (types.isEmpty && schemes.isEmpty) {
+                if (intent.iType.isEmpty && data.isEmpty) {
+                    matchFound(intent, filter)
+                }
+            } else {
+                if (data.nonEmpty) {
+                    val ssi = data.indexOf(":")
+                    val scheme = data.substring(0, ssi)
+                    if (schemes.contains(scheme)) {
+                        val authorities = filter.dataAuthorities
+                        val authorityEnd = findAuthority(data)
+                        if (authorityEnd > ssi + 3) {
+                            val authority = data.substring(ssi + 3, authorityEnd)
+
+                            if (authorities.contains(authority)) {
+                                val paths = filter.dataPaths
+                                if (paths.isEmpty) {
+                                    matchFound(intent, filter)
+                                } else if (ssi < authorityEnd) {
+                                    val path = findPath(data, authorityEnd)
+
+                                    if (paths.exists(p ⇒ p.startsWith(path))) {
+                                        val types = filter.dataTypes
+                                        val iType = intent.iType
+                                        if (types.isEmpty && iType.isEmpty || types.contains(iType)) {
+                                            matchFound(intent, filter)
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    def overApproximateEdges(
```

Comment:
Seems to be used only once, so better not have a separate method for such a small functionality.